### PR TITLE
[Core] Fix core worker client pool leak

### LINKED ISSUE #41260: Ray remote task + fastapi memory leak
### What happened + What you expected to happen

I started a fastapi app and an interface continuously created remote tasks.

I use `memray` tool get memory leak report:

After 1w+ request "/tasks": 
0x1	56.3 MiB	malloc 1239346	 invocation at /usr/local/lib/python3.10/site-packages/ray/remote_function.py:411

After 2w+ request "/tasks": 
0x27	150.3 MiB	malloc	3198029	invocation at /usr/local/lib/python3.10/site-packages/ray/remote_function.py:411

image

image

### Versions / Dependencies

ray version == 2.7.1, and ray cluster run in docker.

### Reproduction script

Here is my fastapi.py
```
@ray.remote
def run_task(entrypoint, tid, metadata):
    entrypoint += f" --job='{json.dumps(metadata)}'"
    cmd = shlex.split(entrypoint)
    logging.info(f"run {cmd}")
    with Popen(args=cmd) as p:
        ...

@app.post("/tasks/")
async def fop_jobs(req: FopReq):
    a = run_fop.options(
        num_cpus=req.entrypoint_num_cpus,
        num_gpus=req.entrypoint_num_gpus,
        runtime_env=req.runtime_env
    ).remote(req.entrypoint, req.tid, req.metadata)
    return req

if __name__ == "__main__":
    uvicorn.run(app, host='0.0.0.0', port=8000)

```

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Currently core worker client pool doesn't remove clients in most cases (there are one or two places where `Disconnect()` might be called) and this caused memory leak. This PR adds a GC inside core worker client pool to remove IDLE clients (i.e.g clients that don't have active connections).

## Related issue number
Closes #41260

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -781,6 +781,8 @@ RAY_CONFIG(int64_t, grpc_client_keepalive_time_ms, 300000)
 /// grpc keepalive timeout for client.
 RAY_CONFIG(int64_t, grpc_client_keepalive_timeout_ms, 120000)
 
+RAY_CONFIG(int64_t, grpc_client_idle_timeout_ms, 1800000)
```
Comment (jjyao) [inline src/ray/common/ray_config_def.h:784]: This is the grpc default value: 30 minutes.

File: src/ray/rpc/grpc_client.h [hunk 2]
```diff
@@ -156,10 +156,20 @@ class GrpcClient {
...
 
   std::shared_ptr<grpc::Channel> Channel() const { return channel_; }
 
+  /// A channel is IDLE when it's first created before making any RPCs
+  /// or after GRPC_ARG_CLIENT_IDLE_TIMEOUT_MS of no activities since the last RPC.
+  /// This method detects IDLE in the second case.
+  /// Also see https://grpc.github.io/grpc/core/md_doc_connectivity-semantics-and-api.html
+  /// for channel connectivity state machine.
+  bool IsChannelIdleAfterRPCs() const {
+    return (channel_->GetState(false) == GRPC_CHANNEL_IDLE) && call_method_invoked_;
```
Comment (rkooo567) [inline src/ray/rpc/grpc_client.h:170]: GetState is not blocking right?

**Review by rkooo567 — APPROVED** [review]

File: src/ray/rpc/worker/core_worker_client_pool.cc [hunk 3]
```diff
@@ -17,31 +17,47 @@
...
+
+  RemoveIdleClients();
+
+  CoreWorkerClientEntry entry;
   auto id = WorkerID::FromBinary(addr_proto.worker_id());
   auto it = client_map_.find(id);
   if (it != client_map_.end()) {
-    return it->second;
+    entry = *it->second;
+    client_list_.erase(it->second);
```
[the remark this answers is not in the document]
Comment (jjyao) [reply src/ray/rpc/worker/core_worker_client_pool.cc:32]: std::list is doubly linked list so it's constant time.

File: src/ray/rpc/grpc_client.h (hunk 2 above)
Comment (jjyao) [reply src/ray/rpc/grpc_client.h:170]: Not blocking.

**Comment by m-harmonic** [comment]
Hello, am I understanding correctly that this fix is not yet merged into any release? Thanks

**Comment by jjyao** [comment]
@m-harmonic Yes, it will be part of Ray 2.10 release.

**Pull request merged**